File: merge-sort/merge_sort.cs
using System;
using System.Linq;

class MergeSort {
    static int[] Sort(int[] arr) {
        if (arr.Length <= 1) return arr;
        
        int mid = arr.Length / 2;
        var left = Sort(arr.Take(mid).ToArray());
        var right = Sort(arr.Skip(mid).ToArray());
        
        return Merge(left, right);
    }
    
    static int[] Merge(int[] left, int[] right) {
        var result = new int[left.Length + right.Length];
        int i = 0, j = 0, k = 0;
        
        while (i < left.Length && j < right.Length) {
            if (left[i] <= right[j]) {
                result[k++] = left[i++];
            } else {
                result[k++] = right[j++];
            }
        }
        
        while (i < left.Length) result[k++] = left[i++];
        while (j < right.Length) result[k++] = right[j++];
        
        return result;
    }
    
    static void Main() {
        int[] arr = {64, 34, 25, 12, 22, 11, 90};
        Console.WriteLine(string.Join(" ", Sort(arr)));
    }
}

File: merge-sort/merge_sort.rs
fn merge_sort(mut arr: Vec<i32>) -> Vec<i32> {
    if arr.len() <= 1 { return arr; }
    
    let mid = arr.len() / 2;
    let right = arr.split_off(mid);
    
    merge(merge_sort(arr), merge_sort(right))
}

fn merge(left: Vec<i32>, right: Vec<i32>) -> Vec<i32> {
    let mut result = Vec::new();
    let (mut i, mut j) = (0, 0);
    
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            result.push(left[i]);
            i += 1;
        } else {
            result.push(right[j]);
            j += 1;
        }
    }
    
    result.extend_from_slice(&left[i..]);
    result.extend_from_slice(&right[j..]);
    result
}

fn main() {
    let arr = vec![64, 34, 25, 12, 22, 11, 90];
    println!("{:?}", merge_sort(arr));
}
